src/hotspot/share/runtime/monitorDeflationThread.cpp

```diff
@@ -0,0 +1,98 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "classfile/javaClasses.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/java.hpp"
+#include "runtime/javaCalls.hpp"
+#include "runtime/monitorDeflationThread.hpp"
+#include "runtime/mutexLocker.hpp"
+
+MonitorDeflationThread* MonitorDeflationThread::_instance = NULL;
+
+void MonitorDeflationThread::initialize() {
+  EXCEPTION_MARK;
+
+  const char* name = "Monitor Deflation Thread";
+  Handle string = java_lang_String::create_from_str(name, CHECK);
+
+  // Initialize thread_oop to put it into the system threadGroup
+  Handle thread_group (THREAD, Universe::system_thread_group());
+  Handle thread_oop = JavaCalls::construct_new_instance(
+                          SystemDictionary::Thread_klass(),
+                          vmSymbols::threadgroup_string_void_signature(),
+                          thread_group,
+                          string,
+                          CHECK);
+
+  {
+    MutexLocker mu(THREAD, Threads_lock);
+    MonitorDeflationThread* thread =  new MonitorDeflationThread(&monitor_deflation_thread_entry);
+
+    // At this point it may be possible that no osthread was created for the
+    // JavaThread due to lack of memory. We would have to throw an exception
+    // in that case. However, since this must work and we do not allow
+    // exceptions anyway, check and abort if this fails.
+    if (thread == NULL || thread->osthread() == NULL) {
+      vm_exit_during_initialization("java.lang.OutOfMemoryError",
+                                    os::native_thread_creation_failed_msg());
+    }
+
+    java_lang_Thread::set_thread(thread_oop(), thread);
+    java_lang_Thread::set_priority(thread_oop(), NearMaxPriority);
+    java_lang_Thread::set_daemon(thread_oop());
+    thread->set_threadObj(thread_oop());
+    _instance = thread;
+
+    Threads::add(thread);
+    Thread::start(thread);
+  }
+}
+
+void MonitorDeflationThread::monitor_deflation_thread_entry(JavaThread* jt, TRAPS) {
+  while (true) {
+    {
+      // Need state transition ThreadBlockInVM so that this thread
+      // will be handled by safepoint correctly when this thread is
+      // notified at a safepoint.
+
+      // This ThreadBlockInVM object is not also considered to be
+      // suspend-equivalent because MonitorDeflationThread is not
+      // visible to external suspension.
+
+      ThreadBlockInVM tbivm(jt);
```

Comment:
Ummm... we use a JavaThread because we have to stop async deflation
during safepoints so that we're not messing with Object headers during GC.
Yes, it's possible to use a non-JavaThread because this is "just software",
but I don't want to try to figure out those races...

Reply:
Ok, I see. This would be bad if it was not a JavaThread.

Reply:
Ok, I see why it has to be a JavaThread.